File: type-parser/mini-parser-ts/test/check_unsupported.ts
#!/usr/bin/env node
/// Assert that the parser rejects the deliberately-unsupported types.
/// A TypeScript port of the Python `check_unsupported.py`.
///
/// Usage: tsx test/check_unsupported.ts [--cases test/cases_unsupported.txt]

import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

import { parseDataType } from "../src/index.js";
import { readCases } from "./cases.js";

const here = dirname(fileURLToPath(import.meta.url));

function main(): number {
  const argv = process.argv.slice(2);
  let casesPath = join(here, "cases_unsupported.txt");
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--cases") casesPath = argv[++i] ?? casesPath;
  }

  const cases = readCases(casesPath);
  let failures = 0;
  for (const typeStr of cases) {
    if (!parseDataType(typeStr).ok()) {
      console.log(`  ok  rejected: ${typeStr}`);
    } else {
      failures++;
      console.log(`FAIL  unexpectedly accepted: ${typeStr}`);
    }
  }

  console.log(`\n${cases.length - failures}/${cases.length} correctly rejected`);
  return failures ? 1 : 0;
}

process.exit(main());

File: type-parser/mini-parser-extracted/test/check_unsupported.py
#!/usr/bin/env python3
"""Assert that chdt-parse rejects the deliberately-unsupported types."""

import argparse
import subprocess
import sys


def read_cases(path):
    cases = []
    with open(path) as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith("#"):
                cases.append(line)
    return cases


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--tool", required=True)
    ap.add_argument("--cases", required=True)
    args = ap.parse_args()

    cases = read_cases(args.cases)
    failures = 0
    for type_str in cases:
        out = subprocess.run([args.tool, type_str], capture_output=True, text=True)
        if out.returncode != 0:
            print(f"  ok  rejected: {type_str}")
        else:
            failures += 1
            print(f"FAIL  unexpectedly accepted: {type_str}")

    print(f"\n{len(cases) - failures}/{len(cases)} correctly rejected")
    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(main())
